### PR TITLE
Transferring all array variable components to different sub-apps in sample transfer

### PR DESCRIPTION
**Pull request by zachmprince**
Refs #20818

### CODE REVIEW

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C [hunk 1]
```diff
@@ -42,6 +42,13 @@ MultiAppVariableValueSamplePostprocessorTransfer::validParams()
       "This should most likely be a Reciever Postprocessor.");
   params.addRequiredParam<VariableName>("source_variable", "The variable to transfer from.");
   params.addParam<unsigned int>("source_variable_component", 0, "The component of source variable");
```
Comment (GiudGiud) [inline framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:44]:
```suggestion
  params.addParam<unsigned int>("source_variable_component", 0, "The component of source variable, may be non-zero for array variables");
```

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C [hunk 2]
```diff
@@ -115,34 +131,39 @@ MultiAppVariableValueSamplePostprocessorTransfer::cacheElemToPostprocessorData()
...
+              "'. The code chooses the sub-application with the smallest ID to set "
+              "the variable on the element, which may created undesired variable solutions."
+              "\nHaving different positions for sub-applications, "
+              "a centroid-based MultiApp or adding block restriction to the variable can "
+              "be used to resolve this warning.");
+
+        // Note: in case of count>0, the sub-application with smallest id will be used for the
+        //       transfer.
+        _cached_multiapp_pos_ids.push_back(multiapp_pos_id);
+        _needed_postprocessors.insert(multiapp_pos_id);
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:165]: we should really only do this once in the new mode, then index into these arrays at the unique location

File: framework/include/transfers/MultiAppVariableValueSamplePostprocessorTransfer.h [hunk 3]
```diff
@@ -51,6 +51,10 @@ class MultiAppVariableValueSamplePostprocessorTransfer : public MultiAppTransfer
   unsigned int _comp;
   /// the moose variable
   MooseVariableFieldBase & _var;
+  /// Whether or not to use array variable to components to separate sub-applications
```
Comment (GiudGiud) [inline framework/include/transfers/MultiAppVariableValueSamplePostprocessorTransfer.h:54]:
```suggestion
  /// Whether or not to map groups of child applications to each component of an array variable
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
I cant say I have the answer on the crucial topics of how everything should be vectorized, at how many levels deep

Seems workable this way.
If you wanted to separate in a second object instead, I d call it
MultiAppArrayVariableComponentSamplePostprocessorTransfer

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C [hunk 4]
```diff
@@ -115,34 +131,39 @@ MultiAppVariableValueSamplePostprocessorTransfer::cacheElemToPostprocessorData()
...
+        for (unsigned int c = 0; c < _apps_per_component; ++c, ++j)
         {
-          distance = current_distance;
-          multiapp_pos_id = j;
-          count = 0;
+          Real current_distance = (getFromMultiApp()->position(j) - elem->true_centroid()).norm();
+          if (MooseUtils::absoluteFuzzyLessThan(current_distance, distance))
+          {
+            distance = current_distance;
+            multiapp_pos_id = j;
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:145]: It's incremented in this loop:
```C++
        for (unsigned int c = 0; c < _apps_per_component; ++c, ++j)
```

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C (hunk 2 above)
[the remark this answers is not in the document]
Comment (zachmprince) [reply framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:165]: Not sure what you mean. Theoretically, you can have different locations for each component (at least that's how I wrote it). To support this, `_cached_multiapp_pos_ids` must be n_elems x n_components.

File: test/tests/transfers/multiapp_variable_value_sample_transfer/tests [hunk 5]
```diff
@@ -21,6 +21,37 @@
                   "point to post processing variable on the sub-application."
   []
 
+  [array_sample_test]
+    type = Exodiff
+    input = parent_array_sample.i
+    exodiff = parent_array_sample_out.e
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply test/tests/transfers/multiapp_variable_value_sample_transfer/tests:27]: It covers both. The answer from the sub app depends on what is sent to it.

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C [hunk 6]
```diff
@@ -316,8 +351,13 @@ MultiAppVariableValueSamplePostprocessorTransfer::execute()
                       "The variable must be a constant monomial with one DoF on an element");
           mooseAssert(pp_values[_cached_multiapp_pos_ids[i]] != std::numeric_limits<Real>::max(),
                       "We should have pulled all the data we needed.");
-          solution.set(dof_indices[0] + _comp, pp_values[_cached_multiapp_pos_ids[i]]);
-          ++i;
+          for (unsigned int c = 0; c < n_subapps / _apps_per_component; ++c)
+          {
+            unsigned int comp =
+                _use_array_sampling ? _cached_multiapp_pos_ids[i] / _apps_per_component : _comp;
```
[the remark this answers is not in the document]
Comment (zachmprince) [reply framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:357]: Nevermind, it's not that different. I will make that function.

File: framework/doc/content/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.md [hunk 7]
```diff
@@ -5,7 +5,7 @@
 ## Description
 
 This transfer supports data transfers between a variable on the main application and a postprocessor on the sub-applications.
-The variable can be a component of an array variable or a standard field variable.
+The variable can be a standard field variable, component of an array variable, or all components of an array variable.
```
Comment (GiudGiud) [inline framework/doc/content/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.md:8]:
```suggestion
The variable can be a standard field variable, a component of an array variable, or all components of an array variable.
```

File: framework/doc/content/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.md [hunk 8]
```diff
@@ -25,6 +30,10 @@ at the sub-application positions to a Postprocessor on each sub-application.
 
 !listing centroid_multiapp/centroid_multiapp.i block=Transfers
 
+An example showing how to transfer all the components of an array variable is shown below. Note that the source variables have 2 components and the MultiApp has two sets of 4 positions.
```
Comment (GiudGiud) [inline framework/doc/content/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.md:33]:
```suggestion
An example showing how to transfer all the components of an array variable is shown below. Note that the source variables have 2 components and the MultiApp has two sets of 4 positions, so 8 child apps. The four first child application, tied to the first set of positions, receive the first variable component (at index 0). The second set of child applications receive the second variable component.
```

Comment (GiudGiud) [reply framework/doc/content/source/transfers/MultiAppVariableValueSamplePostprocessorTransfer.md:33]: I guess it s duplicated with the parameter but oh well

File: test/tests/transfers/multiapp_variable_value_sample_transfer/parent_array_sample.i [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    to_multi_app = sub
+    postprocessor = from_parent
+    source_variable = u_parent
+    map_array_variable_components_to_child_apps = true
+  []
+  [from_transfer]
+    type = MultiAppVariableValueSamplePostprocessorTransfer
+    from_multi_app = sub
+    postprocessor = to_parent
+    source_variable = u_sub
```
Comment (GiudGiud) [inline test/tests/transfers/multiapp_variable_value_sample_transfer/parent_array_sample.i:75]: so this is called source variable but we are setting it with the postprocessor?

Comment (zachmprince) [reply test/tests/transfers/multiapp_variable_value_sample_transfer/parent_array_sample.i:75]: Correct.

Comment (GiudGiud) [reply test/tests/transfers/multiapp_variable_value_sample_transfer/parent_array_sample.i:75]: well it s not your fault but it s not great

**Review by GiudGiud — APPROVED** [review]

File: framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C (hunk 4 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/transfers/MultiAppVariableValueSamplePostprocessorTransfer.C:152]: mad tricks

**Pull request merged**